lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,260 @@
+
+# NOTE: We won't always be able to import from snowflake.{connector, snowpark}.* so need
+# the `type: ignore` comment below, but that comment will explode if `warn-unused-ignores`
+# is turned on when the package is available. Unfortunately, mypy doesn't provide a good
+# way to configure this at a per-line level :(
+# mypy: no-warn-unused-ignores
+
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, Dict, Sequence, cast
+
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.connections.util import load_from_snowsql_config_file, running_in_sis
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.connector import (  # type: ignore # isort: skip
+        SnowflakeConnection as InternalSnowflakeConnection,
+    )
+    from snowflake.connector.cursor import SnowflakeCursor  # type: ignore
+    from snowflake.snowpark.session import Session  # type: ignore
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account"}
+
+
+def _validate_connection_params(params: Dict[str, Any]) -> None:
+    for p in _REQUIRED_CONNECTION_PARAMS:
+        if p not in params:
+            raise StreamlitAPIException(f"Missing Snowflake connection param: {p}")
+
+
+class SnowflakeConnection(BaseConnection["InternalSnowflakeConnection"]):
+    """A connection to Snowflake using the Snowflake Python Connector. Initialize using
+    ``st.connection("<name>", type="snowflake")``.
+
+    SnowflakeConnection supports direct SQL querying using ``.query("...")``, access to
+    the underlying Snowflake Python Connector object with ``.raw_connection``, and other
+    convenience functions. See the methods below for more information.
+    SnowflakeConnections should always be created using ``st.connection()``, **not**
+    initialized directly.
+    """
+
+    def _connect(self, **kwargs) -> "InternalSnowflakeConnection":
+        import snowflake.connector  # type: ignore
+        from snowflake.snowpark.context import get_active_session  # type: ignore
+
+        # If we're running in SiS, just call get_active_session() and retrieve the
+        # lower-level connection from it.
+        if running_in_sis():
+            session = get_active_session()
+
+            if hasattr(session, "connection"):
+                return session.connection
+            return session._conn._conn
+
+        # Otherwise, attempt to create a new connection from whatever credentials we
+        # have available.
+
+        st_secrets = self._secrets.to_dict()
+        if len(st_secrets):
+            # TODO(vdonato): Double-check that we want to merge any secrets we find with
+
+            # the kwargs passed to `st.connection`.
+            conn_kwargs = {**st_secrets, **kwargs}
+            _validate_connection_params(conn_kwargs)
+            return snowflake.connector.connect(**conn_kwargs)
```

Comment:
It would be nice if I can do this:

```toml
# .streamlit/secrets.toml

[connections.snowflake]
connection_name = "my-connection-name-from-snowcli"
```

I *think* it would just need to pass the validation for that to work. Right now it fails with `StreamlitAPIException: Missing Snowflake connection param: account`. Might be worth adding an exception, WDYT?

Reply:
Hm, I'll punt on the decision for this one until Monday (I'm out early in the afternoon tomorrow) / potentially in a small followup PR. This seems reasonable to me, but if we have to implement logic to accept either a `connection_name` or an `account`, then I think it may be more reasonable to just not validate any parameters and allow the underlying library to throw the error in case a user did something wrong.

This won't be quite as nice as the more friendly error message that we provide right now, but the amount of extra handling to implement something that already exists one layer down may not be worth it for anything more complex than very simple error checks